Introduce an asynchronous lookup API

nresare opened the pull request:
I have started experimenting with an asynchronous lookup mechanism. Please note that this is a draft and not not feature complete. I just want to give the opportunity to gather some early feedback. In particular I am curious to to hear about:

* The name `DNSSession` for the new long lived API endpoint. Session is kind of vague on purpose, to allow it to have other public methods in the future that interact with DNS servers. I want to avoid Session because of the inconvenience of having it clash with for example the Session classes in the Neo4j or Cassandra.
* What do we think about putting an exception hierarchy in a separate module? My main reason for going there is that the org.xbill.DNS module is pretty crowded, but this is a weak preference.

The idea is to implement the following:
* Add aliases to QueryResult
* caching
* search path expansion
* ndots handling
* the cycleResults feature
* a configurable minimum credibility requirement
Once this is in place, this query implementation would reach feature parity with Lookup.run() and we could replace that one with this without changing the public API (besides making it Deprecated, possibly).

ibauersachs commented:
> I have started experimenting with an asynchronous lookup mechanism. Please note that this is a draft and not not feature complete. I just want to give the opportunity to gather some early feedback. In particular I am curious to to hear about:
> 
> * The name `DNSSession` for the new long lived API endpoint. Session is kind of vague on purpose, to allow it to have other public methods in the future that interact with DNS servers. I want to avoid Session because of the inconvenience of having it clash with for example the Session classes in the Neo4j or Cassandra.

The prefix is goog, but please try to avoid the uppercasing for new stuff, i.e. use `DnsSession`. I don't have other thoughts yet.

> * What do we think about putting an exception hierarchy in a separate module? My main reason for going there is that the org.xbill.DNS module is pretty crowded, but this is a weak preference.

A separate package for all things lookup, e.g. org.xbill.DNS.lookup is fine, but placing exceptions in a dedicated package is IMO weird. `IOException` for example is in `java.io`, not `java.io.exceptions`.

ibauersachs commented on src/main/java/org/xbill/DNS/DNSSession.java:
```diff
@@ -0,0 +1,129 @@
+package org.xbill.DNS;
```
Please add `// SPDX-License-Identifier: BSD-2-Clause` to new files to clearly indicate their license.

nresare replied:
Sure. How do we feel about adding copyright statements? 

I guess the main purpose for such things would be if someone would make an effort to contact all copyright holders to investigate the viability of a license change in the future. I am happy with or without, and I'm happy with 2 clause BSD.

ibauersachs replied:
The Git history is enough to follow the copyright. The Kernel and systemd do it the same with the simple SPDX comment.

ibauersachs commented on src/main/java/org/xbill/DNS/DNSSession.java:
```diff
@@ -0,0 +1,129 @@
+   * Make an asynchronous lookup of the provided name.
+   *
+   * @param name the name to look up.
+   * @param type the type to look up, values should correspond to constants in Type.
```
Link to classes with `{@link Type}` (in other places too)

nresare replied:
Adding..

ibauersachs commented on src/main/java/org/xbill/DNS/DNSSession.java:
```diff
@@ -0,0 +1,129 @@
+
+      if (firstAnswer.getType() == Type.DNAME || firstAnswer.getType() == Type.CNAME) {
+        resolver
+            .sendAsync(Message.newQuery(buildRedirectQuery(response)))
```
I know this is like the original Lookup class. But is it even legal for a non-recursive resolver to follow CNAMEs if the answer from a recursing resolver is incomplete?

nresare replied:
I agree. It is strange. I just wanted to make it a drop in replacement. Come to think of it I think that it makes sense to have maxRedirects default to 0 and have that mean to not follow redirects and have Lookup use the old behaviour at least until the next major release.

ibauersachs replied:
maxRedirects in the existing Lookup class uses that parameter also to follow the chain within a Message, which is fine. Not following CNAMEs by sending more queries to a (recursive) resolvers is something that can be done in a minor version, or changed via an option.

nresare replied:
I'll look into changing that as a followup to this, but I feel I would like to land this first and implementing the old behaviour first seems like a good idea.

ibauersachs commented on src/main/java/org/xbill/DNS/LookupResult.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
You can use [Lombok](https://projectlombok.org/features/EqualsAndHashCode) for this: add `@EqualsAndHashCode` to the class.

nresare replied:
Thank you for reminding me. I have encountered lombok at the margins but never really engaged with it. This seems very useful. It seems if you add `lombok.addLombokGeneratedAnnotation = true` to a `lombok.config` in the toplevel directory it will exclude generated code from the code coverage calculation as well. I'll add that.

ibauersachs replied:
I don't think I can ever code Java without Lombok again :-)

Please make sure the option doesn't introduce a runtime dependency on Lombok via the `@lombok.Generated` annotation.

nresare replied:
Some googling seems to suggest that annotations that byte code parser encountering annotations that are not on the classpath will ignore them, at least after this bug was fixed: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6322301

ibauersachs replied:
Guava has a bunch of transitive dependencies from annotation libraries that shouldn't be necessary. Not quite sure where the issue lies exactly, Maven, Java, an IDE, whatever, but they need to be there that stuff doesn't fail. See e.g. the discussions in google/guava#3302.

nresare replied:
Having taken a brief look at the referenced issues I don't think that these are comparable, but I will verify this locally.

ibauersachs replied:
I may have picked up the wrong issues in Guava, but all the stuff in their dependency chain that comes along with Guava (failureaccess, listenablefuture, jsr305, errorprone, checker-qual, etc.) is all just for static analysis tools but somehow needed at runtime, at least in some situations (that aren't obvious).

ibauersachs commented on src/main/java/org/xbill/DNS/LookupResult.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Override
+  public String toString() {
```
Can also be done with [Lombok](https://projectlombok.org/features/ToString) `@ToString`.

nresare replied:
changing

ibauersachs commented on src/test/java/org/xbill/DNS/DNSSessionTest.java:
```diff
@@ -0,0 +1,225 @@
+    DNSSession dnsSession = new DNSSession(mockResolver);
+    CompletionStage<LookupResult> resultFuture = dnsSession.lookupAsync(name("a.b."), A, IN);
+
+    assertEquals(new LookupResult(emptyList(), NXDOMAIN), resultFuture.toCompletableFuture().get());
```
I'm just not convinced that an empy list is really the best answer for NXDOMAIN, but neither are `Optional`s or exceptions.

Ideally the return type would be tuple of (result, records) so callers would be forced to not ignore it.

ibauersachs replied:
Do you have an idea/examples of how non-unit test sync and async code should look like?

nresare replied:
Let me think some more about how calling code might look and get back to you on that.

nresare replied:
I'll write something as a new comment below, as this is a comment on an outdated piece of code.

ibauersachs commented on src/main/java/org/xbill/DNS/LookupResult.java:
```diff
@@ -0,0 +1,61 @@
+import org.xbill.DNS.exceptions.AdditionalDetail;
+
+/** LookupResult instances holds the result of a successful lookup operation. */
+public class LookupResult {
```
If the intention is that this class is immutable, make it `final`.

nresare replied:
Good catch, updated

ibauersachs commented on src/main/java/org/xbill/DNS/DNSSession.java:
```diff
@@ -0,0 +1,129 @@
+        case Rcode.SERVFAIL:
+          throw new ServerFailedException();
+        default:
+          throw new LookupFailedException(format("Unknown non-success error code %d", rcode));
```
Use `Rcode.string(rcode)` to translate the numeric value. The underlying mnemonic class takes care of handling unknown values, so it's guaranteed to have a sensible message.

nresare replied:
That is great. At some point I think we should discuss moving to proper enums for the mnemonic classes, but that is for a later time.

ibauersachs replied:
Enums: yes and no. A lot of stuff probably should stay as it is, not just because of the API compatibility, but also because DNS is extensible and with the integers everything still works when new stuff is introduced even if we don't update anything.

nresare replied:
Good point!

ibauersachs commented on src/main/java/org/xbill/DNS/DNSSession.java:
```diff
@@ -0,0 +1,129 @@
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+
+  private final Resolver resolver;
+  private volatile int maxRedirects = DEFAULT_MAX_ITERATIONS;
```
Why volatile?

nresare replied:
It will ensure that reads after a write will get the new value even if it is in a different thread. So, if thread A calls `setMaxRedirects()` with a new value and then thread B calls `lookupAsync()` the updated value will be visible to thread B. Without volatile, thread B might hold a cached copy of the old value indefinitely and not ever see the update.

nresare replied:
Since we have moved to a builder pattern now, this point is moot.

ibauersachs commented on src/main/java/org/xbill/DNS/DNSSession.java:
```diff
@@ -0,0 +1,129 @@
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+public class DNSSession {
```
- Session somewhat implies it needs closing or that it expires, not sure this really applies here
- The async stuff creates some interesting threading issues once there are more config options than a simple integer. E.g. what if there's a cache setter, and the cache is changed while an async lookup is still in progress?
- The existing Cache class is not thread safe (IIRC), so it cannot be reused

nresare replied:
I like immutable classes, with the drawback that configuration typically would be through a long set of constructor args or via the Builder pattern which is a bit verbose. Let me think some more about this

ibauersachs replied:
A builder seems appropriate - and Lombok to the rescue ;-)

nresare commented:
A separate lookup package seems like a good idea to me. I'm happy to go with `DnsSession` for now, looking at the list of classes it is easy to get the impression that we do appreviations in all caps, but I guess that is just because the RFCs uses that for the different data types.

ibauersachs commented:
> A separate lookup package seems like a good idea to me. I'm happy to go with `DnsSession` for now, looking at the list of classes it is easy to get the impression that we do appreviations in all caps, but I guess that is just because the RFCs uses that for the different data types.

The all-caps is mostly historical and I would e.g. create `DnssecRecord` instead of `DNSSECRecord` today. Except for the Record implementations, I'd go for the more sane version. But I won't insist on it if you prefer `DNSSession`.

ibauersachs commented:
About that "session", yesterday I tought about using "LookupSession". DnsSession might be a bit too generic. Just an idea.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/DnsSession.java:
```diff
@@ -0,0 +1,214 @@
+                        throw new IllegalArgumentException("Search path name too long");
+                      }
+                    })
+                .collect(Collectors.toCollection(ArrayList::new));
```
Use `Collectors.toList()`

nresare replied:
Lombook generates the searchPath property as being of type ArrayList, so the code that makes the names absolute needs to return an ArrayList as well to be able to update the value.

ibauersachs replied:
Ah. But then `.forEach(searchPath::add);` is IMO more better than overwriting an already instantiated field.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/DnsSession.java:
```diff
@@ -0,0 +1,214 @@
+    if (name.length() == 255) {
+      // handle this special case where the name is relative but can not be made absolute as it
+      // is at it's maximum length and nothing can be appended.
+      throw new IllegalArgumentException("Name is too long");
```
Can't you use `NameTooLongException`?

nresare replied:
Since `NameTooLongException` is a checked exception it would add an exception to the `lookupAsync` and one for a pretty esoteric case as well. (Since it takes a Name and names can't be constructed if they are too long, it is only for the case where it is a maximally long relative name that can not be made absolute).

I think this is a good use case for a checked exception, as it specifies exactly what is going on for the edge case but for the common case it can be ignored and the failure will bubble up the call stack. Callers that want to deal with it can do that, but callers that don't doesn't need to.

Moreover, since I envision lookupAsync() to be used in chains of CompletionStage based async code, having it throw a checked exception would make those chains a lot more verbose as it would need to be either handled out of the call stack or wrapped in an unchecked exception.

nresare replied:
Come to think of it I think that the better way to deal with this is to change Name to fail creation for relative names of length 255. That way the caller would need to handle that the same way as it would need to handle other Name creations that are much longer.

ibauersachs replied:
Without following the code too much now, what happens with two search paths where only the second will generate a NTLE? If that is communicated as a failed future, you could do the same with with the first path (by try/catching expandName at the call site and immediately returning a failed future. No need for a checked exception in the signature of lookupAsync.

As for limiting the Name length, not sure. Let's discuss this somwhere else.

nresare replied:
Ah, I did not think of that. Having any checked exception be passed on through the CompletionStage is definitely one way of doing it.

I'm okay with that, although I think that changing the Name constructor to disallow it is the better tradeoff, but we can discuss that in the PR that I just opened. 

My fairly strong opinion is that checked exceptions as designed in Java is a flawed concept, but that is not something that we will fix tonight.

ibauersachs replied:
The concept is broken, yes. Especially now with the lambdas. But Java has been born disabled. Signed bytes. Sigh.

nresare commented:
I'm happy to change the name to LookupSession. If we would want to mix lookups and other operations with resolvers (zone transfers and updates comes to mind) in the future we can always create a new interface that has all of those aspects baked in.

nresare commented:
After having added a few more bits and pieces, I believe that the code is feature complete now. Things left to do:

1. Some more analysis on the pros and cons of the different ways to returning empty non-exceptional results
2. Get rid of the search path too long code path.
3. Add a delombok stage to the build and have javadoc be generated off of that, as the `@Builder` related methods in LookupSession is pretty important to it's functioning

ibauersachs commented:
> I'm happy to change the name to LookupSession.

I don't know if it's really better, just an idea I had while cooking :-)

> If we would want to mix lookups and other operations with resolvers (zone transfers and updates comes to mind) in the future we can always create a new interface that has all of those aspects baked in.

Not sure how you inted to mix that. In any case, I don't see how a zone transfer/update could be logically mixed with the options set in the builder (ndots, search path, cache). I mean okay, you could try to update a relative name, but that just makes no sense at all (and the constructor of Update forbids it).

nresare commented:
> > I'm happy to change the name to LookupSession.
> 
> I don't know if it's really better, just an idea I had while cooking :-)
> 

Having now seen both in front of me, I think that `LookupSession` is a better name.

nresare commented:
So, signalling NXDOMAIN/NXRRSET as a lookup result. I see 5 options in the case of a query resulting in a NXDOMAIN/NXRRSET response from the resolver:

1. The current variant. LookupResult.getResult() returning an empty list.
2. having LookupResult.getResult() return null on NXDOMAIN/NXRRSET.
3. Throwing an exception
4. Having getResult() return an `Optional<List<Record>>` that is empty 
5. have getResult() throw an exception when called

My read is that we would like calling code dealing with this case, and we we would like to nudge callers to take it into account. I think 3-5 does that. 3 has the drawback of using exceptions for a decidedly non-exceptional activity. Take for example code that would make lookups in various DNS based junk email blacklists. Most responses will be NXDOMAIN and although creating Exceptions is not as expensive that that it once was, it still has a cost.

I think that the method signature in 4 is bit too long to look good, but it has the benefit of it being syntactically straightforward to promote the case into an exception (which I assume many callers would want) with `result.getResult().orElseThrow()`.

Come to think of it, if we went with 2, consumers could get that pattern if they wanted with `Optional.of(result.getResult()).orElseThrow()`.

5 has some appeal, as the "high performance" user could look at the additionalDetail field and avoid generating an exception and the casual user that just want NXDOMAIN to be treated as any other failure somewhere up the call stack could get that. However, I am skeptical of diverging from the principle of least astonishment to have a getter throw an exception.

I am leaning towards 2, as it gives a clean way to have Optional.orElseThrow() to get the exception and I can't really think of a case where the empty list would be easier to consume from than handling the null value.

nresare commented:
I rebased and squashed the commits and removed the draft flag, as I think that we are mostly in agreement about this change. Please let me know your thoughts on the NXDOMAIN options I outline above.

nresare commented:
I have now verified that using an artefact built with the lombok annotated byte code (to prevent code coverage from measuring generated bytecode) I can run a project with no lombok dependencies, only dnsjava and there are no issues loading the byte code from that jar. 

```
[INFO] com.resare.maven-test:maven-test:jar:0.0.1
[INFO] +- org.junit.jupiter:junit-jupiter-api:jar:5.7.0:test
[INFO] |  +- org.apiguardian:apiguardian-api:jar:1.1.0:test
[INFO] |  +- org.opentest4j:opentest4j:jar:1.2.0:test
[INFO] |  \- org.junit.platform:junit-platform-commons:jar:1.7.0:test
[INFO] +- org.junit.jupiter:junit-jupiter-engine:jar:5.7.0:test
[INFO] |  \- org.junit.platform:junit-platform-engine:jar:1.7.0:test
[INFO] \- dnsjava:dnsjava:jar:3.4.0-SNAPSOT:compile
[INFO]    \- org.slf4j:slf4j-api:jar:1.7.30:compile
```

ibauersachs commented:
> So, signalling NXDOMAIN/NXRRSET as a lookup result. I see 5 options in the case of a query resulting in a NXDOMAIN/NXRRSET response from the resolver:
> 
> 1. The current variant. LookupResult.getResult() returning an empty list.

I checked out this branch to code an example (based on some code I had lying around) and immediately fell into the trap of expecting an answer because I used the name before, but it was now NXDOMAIN. The empy list thus was surprising.

> 2. having LookupResult.getResult() return null on NXDOMAIN/NXRRSET.

Nah, null for Lists is unexpected.

> 3. Throwing an exception

I don't think this is so bad. In this flow I defintely expect to have a valid, usable answer or an exception, not something in between:

```java
    SimpleResolver sr = new SimpleResolver("8.8.8.8");
    LookupSession l = LookupSession.builder().resolver(sr).ndots(1).maxRedirects(16).build();
    l.lookupAsync(Name.fromConstantString("macmini.local"), Type.A, DClass.IN)
        .toCompletableFuture()
        .thenAccept(
            res -> {
              for (Record r : res.getRecords()) {
                System.out.println(r);
              }
            })
        .exceptionally(
            e -> {
              System.out.println(e.getMessage());
              return null;
            })
        .get();
```

> 4. Having getResult() return an `Optional<List<Record>>` that is empty

Maybe, not sure.

> 5. have getResult() throw an exception when called

Definitely not an option.

> 
> My read is that we would like calling code dealing with this case, and we we would like to nudge callers to take it into account. I think 3-5 does that. 3 has the drawback of using exceptions for a decidedly non-exceptional activity. Take for example code that would make lookups in various DNS based junk email blacklists.

FWIW, the blacklists always return an answer, the status is encoded in the last octet of a 127.0.0.0/8 address.

> Most responses will be NXDOMAIN and although creating Exceptions is not as expensive that that it once was, it still has a cost.

It's bad style to rely on exceptions for expected outcomes. BUT in an async API I'm not so sure anymore. Plus the exception isn't really thrown anymore, so it's less of overhead than an actually thrown exception (but still more than enum).

> I think that the method signature in 4 is bit too long to look good, but it has the benefit of it being syntactically straightforward to promote the case into an exception (which I assume many callers would want) with `result.getResult().orElseThrow()`.

Something in between could be an option (configurable via builder) to let the consumer decide if NX* should result in a failed or succeeded future. Not sure if it's a good idea either.

> 
> Come to think of it, if we went with 2, consumers could get that pattern if they wanted with `Optional.of(result.getResult()).orElseThrow()`.
> 
> 5 has some appeal, as the "high performance" user could look at the additionalDetail field and avoid generating an exception and the casual user that just want NXDOMAIN to be treated as any other failure somewhere up the call stack could get that. However, I am skeptical of diverging from the principle of least astonishment to have a getter throw an exception.
> 
> I am leaning towards 2, as it gives a clean way to have Optional.orElseThrow() to get the exception and I can't really think of a case where the empty list would be easier to consume from than handling the null value.

I haven't seen Optionals used like this.

ibauersachs commented:
> I have now verified that using an artefact built with the lombok annotated byte code (to prevent code coverage from measuring generated bytecode) I can run a project with no lombok dependencies, only dnsjava and there are no issues loading the byte code from that jar.
> 
> ```
> [INFO] com.resare.maven-test:maven-test:jar:0.0.1
> [INFO] +- org.junit.jupiter:junit-jupiter-api:jar:5.7.0:test
> [INFO] |  +- org.apiguardian:apiguardian-api:jar:1.1.0:test
> [INFO] |  +- org.opentest4j:opentest4j:jar:1.2.0:test
> [INFO] |  \- org.junit.platform:junit-platform-commons:jar:1.7.0:test
> [INFO] +- org.junit.jupiter:junit-jupiter-engine:jar:5.7.0:test
> [INFO] |  \- org.junit.platform:junit-platform-engine:jar:1.7.0:test
> [INFO] \- dnsjava:dnsjava:jar:3.4.0-SNAPSOT:compile
> [INFO]    \- org.slf4j:slf4j-api:jar:1.7.30:compile
> ```

Okay, but the annotations need to go on the first sign of trouble.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/AdditionalDetail.java:
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: BSD-2-Clause
+package org.xbill.DNS.lookup;
+
+public enum AdditionalDetail {
```
- Missing class-level Javadoc
- Needs a SUCESSFUL (or OK) case
- A better name might be `LookupResultDetail`

nresare replied:
If we decide to go with exceptions instead we get rid of this whole enum

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/InvalidZoneDataException.java:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: BSD-2-Clause
+package org.xbill.DNS.lookup;
+
+public class InvalidZoneDataException extends LookupFailedException {
```
class-level Javadoc

nresare replied:
Added.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupResult.java:
```diff
@@ -0,0 +1,40 @@
+public final class LookupResult {
+  /** An unmodifiable list of records that this instance wraps, may not be null but can be empty */
+  @Getter private final List<Record> records;
+  /** provides additional detail when records is empty */
```
- Capitalize the beginning
- Use {@link #getRecords()} to refer to the list

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupResult.java:
```diff
@@ -0,0 +1,40 @@
+  @Getter private final List<Record> records;
+  /** provides additional detail when records is empty */
+  @Getter private final AdditionalDetail additionalDetail;
+  /** tracks the different CNAME and DNAME aliases for this name */
```
Capitalize

nresare replied:
Added some more detail on the semantics, and capitalised.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupResult.java:
```diff
@@ -0,0 +1,40 @@
+   * @param additionalDetail additional detail on this response, such as the reason for records
+   *     being null.
+   */
+  public LookupResult(List<Record> records, AdditionalDetail additionalDetail, List<Name> aliases) {
```
I'd keep this package-private as long as there's no need to have it public

nresare replied:
I feel fairly strongly that this should be public because it simplifies testing. If a user of the library want to unit test their code it would be natural to mock a LookupSession. If the test code can not create `LookupResult` instances without resorting to reflection magic (or mocking the actual value object) that adds a lot of complexity to the test setup.

This is something that Apache Kafka suffers from, where the admin api returns value objects without public constructors, and it has been pretty annoying.

ibauersachs replied:
Well, okay, but then please keep in mind that it becomes public API that cannot be changed anymore.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupResult.java:
```diff
@@ -0,0 +1,40 @@
+      throw new IllegalArgumentException(
+          "Empty records can not be combined with null additionalDetail");
+    }
+    this.additionalDetail = additionalDetail;
```
Guarantee that this is not null with the OK status (as Java cannot (yet) handle null in switch statements).

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+import org.xbill.DNS.Type;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
```
Old name used, refers to setters instead of builder

nresare replied:
Thanks! Updated.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}.
```
`... Defaults to {@value #DEFAULT_MAX_ITERATIONS}.`

nresare replied:
For the person reading that comment, I think we should skip the indirection.

nresare replied:
ooh, now I thing I'm getting  `@value` tags. Seems useful, but I can't get it to work for some reason.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+ */
+@Builder(buildMethodName = "innerBuild")
+public class LookupSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
```
Might be worth it having it public. Not sure.

nresare replied:
I think it makes sense to keep it private as long as we are not using it anywhere else.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
+   * absolute. The default is {@code 1}, meaning meaning that if there are any dots in a name, the
```
`{@value #NDOTS}`

Should this default constant be shared with the default in `org.xbill.DNS.config.ResolverConfigProvider`? And I've forgotten how this default interacts with the one from the ResolverConfig anyway.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+                    })
+                .collect(Collectors.toCollection(ArrayList::new));
+      }
+      return innerBuild();
```
The innerBuild method is public. It needs to be at least package-private (found this: https://rusyasoft.github.io/java,%20lombok/2020/02/14/Overriding-build-setter-lombok-builder/, but maybe there's a better way).

nresare replied:
Oh, that was a solution I did not find in my own search. It seemed to me like the it is only possible to control the access level of all the generated method, not specifically the generated builder. But I will use the anonymous subclass solution instead.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+   * @param dclass the class to look up, values should correspond to constants in {@link DClass}.
+   * @return A {@link CompletionStage} what will yield the eventual lookup result.
+   */
+  public CompletionStage<LookupResult> lookupAsync(Name name, int type, int dclass) {
```
add an overload without DClass that defaults to DClass.IN, and maybe one that takes a String instead of a Name

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+    if (name.isAbsolute()) {
+      return Stream.of(name);
+    }
+    if (name.length() == 255) {
```
This is superfluos with #165, right?

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+   * Configures the Cache instances to be used for lookups for the different {@link DClass} values.
+   */
+  @Singular("cache")
+  private final Map<Integer, Cache> caches;
```
Is there a default Cache as before? How can I set one without bothering about what a DClass is?

nresare replied:
This is not yet implemented. I'm thinking that there should be a static method that would give you a _default_ LookupSession with sensible defaults set, but with ~100 comments to this PR, I would prefer to have that land a bit later.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+  @Builder.Default private final int ndots = DEFAULT_NDOTS;
+
+  /** Configures the search path used to look up relative names with less than ndots dots. */
+  @Singular("searchPathElement")
```
This name confused me. Either make it searchPaths/searchPath or both searchPath

nresare replied:
I'm happy to change to have both be searchPath (yay for method overloading!). I don't think that plural searchPaths makes sense as other examples such as `$PATH` has multiple elements but still is a singular thing.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+  private final List<Name> searchPath;
+
+  /**
+   * If set to true, cached results with multiple records will be returned with the starting point
```
`{@code true}`

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+   */
+  @Builder.Default private final int ndots = DEFAULT_NDOTS;
+
+  /** Configures the search path used to look up relative names with less than ndots dots. */
```
`{@link #ndots()}`

nresare replied:
That doesn't seem to work. At least it shows up as red for me in my ide. I'll make a note to look at that when I get around to fixing the javadoc to include lombok generated code.

ibauersachs replied:
That's probably Lombok, `{@link #$default$ndots()}` exists (but makes no sense to link to).

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+   * @return A {@link CompletionStage} what will yield the eventual lookup result.
+   */
+  public CompletionStage<LookupResult> lookupAsync(Name name, int type, int dclass) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
```
Create this future closer to its usage to avoid creation in case of failure

nresare replied:
I believe this is the closest that the future can be created. The construct of having an outer future and then a function doing some asynchronous work one or more times by a recursive call requires the outer future to be defined in the outer function. Also note that It's not necessarily call stack recursion as the recursive call is made as a result of an asynchronous operation, possibly on some worker thread.

ibauersachs replied:
Yes, this comment was meant to be somewhere else... can't find it anymore.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+
+  private CompletionStage<LookupResult> lookupWithCache(Record queryRecord, List<Name> aliases) {
+
+    Supplier<CompletionStage<LookupResult>> lookupWithResolver =
```
I think it would be easier to read if this were a proper method

nresare replied:
Agreed. This was me simply breaking out some part of a needlessly long chain into a variable, but a function is much better. Changed.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,299 @@
+
+  private CompletionStage<LookupResult> setResponseToMessageFuture(SetResponse setResponse) {
+    if (setResponse.isNXDOMAIN()) {
+      return completedFuture(new LookupResult(emptyList(), AdditionalDetail.NXDOMAIN, null));
```
I'm unused to static imports except for assertions and (sometimes) constants

nresare replied:
While this particular code has changed, for the general case I am reading your statement that you are unused to the practice in the context of a change request as that you would rather me not doing that. While I could definitely make a compromise and remove the static imports, I think that it improves readability. Please tell me more about what you think, why and how strong this opinion is :)

ibauersachs replied:
I think it's confusing (with local method names). I'd refer to some guidelines, but I don't have any at hand. If you're used to it from previous work, then I guess I could live with it. I've just never seen static imports for methods used outside of test code.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/RedirectOverflowException.java:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: BSD-2-Clause
+package org.xbill.DNS.lookup;
+
+public class RedirectOverflowException extends LookupFailedException {
```
class-level Javadoc

nresare replied:
Added

ibauersachs reviewed (changes requested):
More nitpicks

nresare commented:
> 
> I don't think [exceptions] is so bad. In this flow I defintely expect to have a valid, usable answer or an exception, not something in between:
> 

Having now experimented a bit, I can throw and catch 100k exceptions in about ~200ms on my laptop. If someone is hurt by the performance implications of that, I think they might have other problems in their design.

Let me implement an exception based approach and we can discuss further on the basis of that.

ibauersachs commented:
> I have pushed a version that uses two new exceptions instead of empty list and an enum to signal NXDOMAIN/NXRRSET. If I may say so myself, I think that this is a better approach.

Yes, it feels natural now.

> At the same time getting rid of the complexity of handling NameTooLongException for the edge case that we are now handling in Name. Unfortunately we now have some code that is not testable, so coverage is down from 100%, but I have a theory of how to address this with a later change.

If you want to make it an exercise, great, but there's really no need to exaggerate :-)

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,295 @@
+  }
+
+  /** Returns a LookupResult if this response was a non-exceptional empty result, else null. */
+  private static LookupResult buildResult(Message answer, List<Name> aliases) {
```
Consider if you really need to throw the exceptions or if you could just return a failed future (the stacktrace in the exceptions is useless and it would probably half your exception-throwing-testbench-time).

And it could be useful to add the query's Name/Type to the exception; the caller should know it, but if the failure/exception is blindly used for logging it's helpful.

nresare replied:
One could move away from having regular synchronous flow in these methods and instead allow for a mechanism that returned either a value or an exception, but I think that might be the wrong tradeoff as it diverges from a very well established pattern of execution with a simple return value for the successful case and an Exception for the exceptional case, simply to save a ms or two. 

I'm not so sure about adding other potentially useful information to the exception that is also available up the call stack. It makes things a bit simpler to not add it at the very least :)

ibauersachs replied:
setResponseToMessageFuture(...) does such a non-throwing, failed-future flow. But I agree that this is currently more readable.

There were a few issues that requested adding such information to exceptions.

nresare commented:
> If you want to make it an exercise, great, but there's really no need to exaggerate :-)

I am happy to not have coverage when there is a good reason not to, but otherwise I think it is useful to strive for 100, and I think it makes for better code and interfaces as well.

nresare commented:
This PR became a bit unwieldy, but I believe that I have mostly addressed the concerns you raised. Please have another look :)

I made an attempt to have the lombok generated public code be part of the javadoc but it was more difficult than I initially anticipated, so I would prefer to have this land as a separate standalone PR. In the last two years I have become very acquainted with gradle, and doing pretty much anything with maven these days feels like a detour, but that is a discussion for another time.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+   */
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
```
much -> must

nresare replied:
fixed.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
+   * absolute. The default is {@value #DEFAULT_NDOTS}, meaning meaning that if there are any dots in
```
`meaning meaning`

nresare replied:
fixed.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+                        throw new IllegalArgumentException("Search path name too long");
+                      }
+                    })
+                .collect(Collectors.toCollection(ArrayList::new));
```
The comment probably got lost, but what about `searchPath.stream()...().forEach(this.searchPath::add);` instead of ArrayList::new? It would avoid the creation of a never used list.

nresare replied:
Well, calling add() on a list from within a stream over that list causes a `ConcurerntModificationException`. While it is good to not create throwaway values left and right, I think that constructing a new list is okay in this case. Let's call it the least worst option.

ibauersachs replied:
Ah right, I didn't pay attention that this is the same list.

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+import static java.util.concurrent.CompletableFuture.*;
```
Wildcard imports are not allowed in g-j-f. Seems I forgot to bind `checkstyle:check` to the verify phase (and I'd still prefer not using static method imports at all).

nresare replied:
Moved to spelling out the classes instead

ibauersachs commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+    List<Record> answer = response.getRecords();
+    Record firstAnswer = answer.get(0);
+    if (answer.size() != 1) {
+      throw new InvalidZoneDataException("Multiple CNAME RRs not allowed, SEE RFC1034 3.6.2");
```
`see` or is there a reason for writing this uppercase?

nresare replied:
No reason 🤷

ibauersachs reviewed (changes requested):
Some small stuff, and this as tbd for follow ups:
- Delombok for Javadoc in of the builder methods
- ndots from/interaction with ResolverConfig
- max iterations sytem property
- simplified lookupAsync(...) overload(s)
- simplified and/or default Cache
- create() with sensible defaults

nresare commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}. Defaults to {@code 16}.
```
I got the `{@value VALUE}` construct to work. (It turns out that with the latest released java 8 javadoc tool, one need to have the fully qualified field name of the constant field, presumably since the javadoc field is copied to an inner class by lombok).

I agree that it makes sense to add reading of the default to the code that constructs the sensible default instance, along with setting up caches and so on.

nresare commented on src/main/java/org/xbill/DNS/lookup/LookupSession.java:
```diff
@@ -0,0 +1,314 @@
+ */
+@Builder
+public class LookupSession {
+  // remember to update javadoc if this is changed for the following two properties.
```
Fixed.

nresare commented:
Thank you Ingo for putting so much effort into this review. I will try to make the upcoming changes smaller and easier to review :)
